libflux/flux-core/src/semantic/types.rs

```diff
@@ -904,7 +904,11 @@ impl MonoType {
             (MonoType::Error, _) | (_, MonoType::Error) => (),
             // Similarly if both sides are dynamic, there's nothing more to do.
             (MonoType::Dynamic(_), MonoType::Dynamic(_)) => (),
-
+            // In situations where a function receives a dynamic value as an
+            // unbounded parameter, this allows the value to be used like a
+            // record (ie, granting member access).
+            (MonoType::Dynamic(_), MonoType::Record(_))
+            | (MonoType::Record(_), MonoType::Dynamic(_)) => (),
```

Comment:
I don't think this makes sense under the assumption we have discussed with dynamic. This effectively allow `dynamic` to be passed anywhere we expect a record (and vice versa). If we go with a more permissive dynamic where dynamic may be accepted everywhere (and therefore "bleeds out more easily" in the code) I think  this change makes sense (though it would be more general and just do this for all types, effectively.

Reply:
Also, I suppose this is a testcase I should add on the rust side; something to show that dynamic is rejected from a builtin that has a Record in the constraint.

I'll try to add that with the continued work here.

Reply:
Sure enough, a test like the following fails by passing typechecks. I'm still tinkering to try and get it green without spoiling the rest.

```rust
#[test]
fn dynamic_rejected_where_record_expected() {
    test_error_msg! {
        env: map![
            "d" => "dynamic",
            "fn" => "(v: A) => bool where A: Record",
        ],
        src: r#"
        x = fn(v: d)
        "#,
        expect: expect![[r#"tbd"#]]
    }
}
```

Reply:
_I think_ 2efb7affabf93e885b7e10da62d0a983a171cd00 should cover the specific concern of Record/dynamic being interchangeable. For places where the `Record` constraint is used, `dynamic` should now be rejected.

Again, this is mostly guesswork. I only know what the test suite is telling me. If there are more edges to this, I'll probably need more tests to capture the scenario.

Reply:
2efb7af seems fine (though it would be better to have `ToInt`, `ToString`, etc kinds rather than an all, encompassing `Basic` kind).

The `Record <=> Dynamic` is  still wrong though as I understand from what we intended with dynamic

(That is, this test should not compile under the current direction we have for dynamic)
```
fn = (x) => ({ f: x.f0 })
v = fn(x: d)
```

Reply:
Yeah, okay. Unfortunately, I think that test will have to work as-is if we want to be able to send `dynamic` into function args like those in `array.map` and `array.filter`. Maybe I've missed something here.

Part of the my hope for this version of dynamic was to be _more strict at the leaves_ while still _permissive at the branches_, so having a close affinity with `Record` makes sense to me.

The only way I can think to handle things differently would be to add a new kind which is `RecordLike` which both `Record` and perhaps a new `DynamicRecord` could constrain to. We could then offer a `dynamic.asObject` converter, like we have for array. 

Either that or we deny member access entirely and emulate the [API we have for dictionary](https://docs.influxdata.com/flux/v0.x/stdlib/dict/get/) which I find to be very clunky.